### ISSUE TITLE
Assistance converting some code to version 5 please

### ISSUE THREAD

**Issue by morrow95** [amcharts/amcharts5#173]
The code below I believe was part of the version 4 real-time update demo chart.  Any chance someone could assist me converting this over for version 5?

```
// all the below is optional, makes some fancy effects					
// this makes date axis labels to fade out
dateAxis.renderer.labels.template.adapter.add("fillOpacity", function (fillOpacity, target) {
    var dataItem = target.dataItem;
    return dataItem.position;
})

// need to set this, otherwise fillOpacity is not changed and not set
dateAxis.events.on("validated", function () {
    am4core.iter.each(dateAxis.renderer.labels.iterator(), function (label) {
        label.fillOpacity = label.fillOpacity;
    })
})

// this makes date axis labels which are at equal minutes to be rotated
dateAxis.renderer.labels.template.adapter.add("rotation", function (rotation, target) {
    var dataItem = target.dataItem;
    if (dataItem.date && dataItem.date.getTime() == am4core.time.round(new Date(dataItem.date.getTime()), "minute").getTime()) {
        target.verticalCenter = "middle";
        target.horizontalCenter = "left";
        return -90;
    }
    else {
        target.verticalCenter = "bottom";
        target.horizontalCenter = "middle";
        return 0;
    }
})
```

**Comment by martynasma** [comment]
You may find the following tutorials useful.

Adapters:
https://www.amcharts.com/docs/v5/concepts/settings/adapters/

DateAxis:
https://www.amcharts.com/docs/v5/charts/xy-chart/axes/date-axis/

**Comment by morrow95** [comment]
Not really.  I'm not seeing examples for things such as `am4core.iter.each()`, `iterator()`, `am4core.time.round()` in terms of version 5.  I've always had trouble with your docs from day 1 though in all versions.  Not everything has examples and the 'reference' section doesn't contain examples at all.  Pretty much all the charts I use are based off demo's you had for version 4 which aren't exactly the same for version 5 either.

**Comment by martynasma** [comment]
How about this:

```JavaScript
xAxis.get("renderer").labels.each(function(label) {
  console.log(label)
})
```

As for time functions, they are identical to v4:
https://www.amcharts.com/docs/v5/reference/time_module/

**Comment by morrow95** [comment]
```
xAxis.get("renderer").labels.template.adapters.add("opacity", function (opacity, target) {
	console.log('fill',opacity);
	console.log('data', target.dataItem);
    //var dataItem = target.dataItem;
    //return dataItem.position;
})
```

is `target.dataItem.position` in another name/location I'm not seeing or just doesn't exist anymore?

**Comment by martynasma** [comment]
Label's data item does not contain position, which I'm assuming relative position within axis.

You can use axis' `dataItemToPosition()` method:

```JavaScript
xAxis.get("renderer").labels.template.adapters.add("opacity", function (opacity, target) {
  console.log(xAxis.dataItemToPosition(target.dataItem));
  return opacity;
})
```

**Comment by morrow95** [comment]
`Uncaught TypeError: d.dataItemToPosition is not a function`

I'm not seeing `dataItemToPosition()` anywhere in the docs either.

**Comment by morrow95** [comment]
Not sure if this is the best way to go about it, but seems like I found something that replicates the behavior I was after : 

```
xAxis.get("renderer").labels.template.adapters.add("opacity", function (opacity, target) {
	//console.log(target.dataItem);
	if(target.dataItem._settings.value){
		return xAxis.valueToPosition(target.dataItem._settings.value);
	}
	else return 0;
})
```

**Comment by morrow95** [comment]
New problem for last part of my original code.  I am checking if the label is 'on the minute' and if so rotating it.  This works, however, in this particular chart I add data to it every 15 seconds or so.  As data is added the chart 'slides' to the left along with its labels.  When this happens the label that was originally correct and rotated is now a different value, but stays rotated.  It is as if the below only runs on chart load... not after new data is added.

```
// this makes date axis labels which are at equal minutes to be rotated
xAxis.get("renderer").labels.template.adapters.add("rotation", function (rotation, target) {
	//console.log(target.dataItem);
	if( target.dataItem._settings.value && new Date(target.dataItem._settings.value).getSeconds() == 0)  {
		return -90;
	} else return 0;
})
```

**Comment by martynasma** [comment]
> It is as if the below only runs on chart load... not after new data is added.

That is exactly what it is. Updating data does not automatically invalidate all elements of the chart. Only those that need to be invalidated. If the label for X date is already there, the chart is saving resources by not invalidating it completely, just moving it to a new position.

In your case, you probably need to force invalidation manually. Something like this:

```JavaScript
xAxis.get("renderer").labels.each(function(label) {
  label.set("rotation", 0);
})
```

This will force labels to be re-rendered and their adapters to kick in.

**Comment by morrow95** [comment]
That makes sense.  In version 4 there was `chartReg['chart'].invalidateData();` - is there something similar in version 5 for this purpose?  Seems like there would be situations I'd need to manually force invalidation on multiple elements depending on what all is being done with the chart.

**Comment by morrow95** [comment]
```
xAxis.get("renderer").labels.each(function(label) {
  label.set("rotation", 0);
})
```

Isn't forcing the adapters to kick in either.  I've set this code where I would have called `invalidateData()` in version 4 - right after adding new data to the chart.  For debugging I've output some markers with console.log() in it and the adapter in question.  The adapter doesn't get hit.

**Comment by martynasma** [comment]
Would you be able to post your whole chart and data on CodePen/jsFiddle so that we can test?

**Comment by morrow95** [comment]
I have a lot of other stuff going on and polling data from a database with ajax, but this mimics it and shows the same problem.

[https://codepen.io/scubapongo/pen/GRMdqaY](https://codepen.io/scubapongo/pen/GRMdqaY)

**Comment by martynasma** [comment]
Thanks for hanging in. Version 5.0.21 added `markDirtyValues()` method on `Component` objects.

You can now do this:

```JavaScript
xAxis.markDirtyValues();
```

After updates, to force axis labels repainted.

**Comment by morrow95** [comment]
@martynasma Great!  I'll give this a go today.  Thank you!

**Comment by morrow95** [comment]
@martynasma Unless the codepen I listed above isn't using 5.0.21 (looks like it should be using the latest 5.x branch though) this doesn't seem to be solving the issue.

**Comment by martynasma** [comment]
My bad, turns out `markDirtyvalues` did not make it into 5.0.21. It will be included in the next version we're releasing today or coming days: 5.1.0. Sorry!
